main.go

```diff
@@ -216,30 +253,59 @@ func getData(ctx context.Context, filter map[string]string, nametag string) ([]*
 		logrus.Warnf("convertASGsToMetrics failed: %v", err)
 		return nil, err
 	}
-	mu.Lock()
-	globalCache = &cache{
-		Date:    t,
-		Metrics: asg,
-	}
-	mu.Unlock()
 	return asg, nil
 }
 
-func metricsHandler(rsp http.ResponseWriter, req *http.Request) {
-	filter := map[string]string{}
-	for _, item := range strings.Split(opts.Filter, ",") {
-		if !strings.Contains(item, "=") {
-			continue
-		} else {
-			spl := strings.Split(item, "=")
-			filter[spl[0]] = spl[1]
+func errorCountMetrics(counts map[string]int) *dto.MetricFamily {
+	generateGaugeFamily := func(name, help string) *dto.MetricFamily {
+		c := dto.MetricType_COUNTER
+		return &dto.MetricFamily{
+			Name:   aws.String(name),
+			Help:   aws.String(help),
+			Type:   &c,
+			Metric: []*dto.Metric{},
 		}
 	}
 
-	out, err := getData(req.Context(), filter, opts.NameTag)
-	if err != nil {
-		logrus.Warnf("Could not get ASG data: %v", err)
+	errFamily := generateGaugeFamily("aws_asg_request_error_count", "Total number of errors in fetching data")
+
+	for msg, count := range counts {
+		generateMetric := func(v float64) *dto.Metric {
+			lp := &dto.LabelPair{Name: aws.String("error"), Value: aws.String(msg)}
+			return &dto.Metric{
+				Label:   []*dto.LabelPair{lp},
+				Counter: &dto.Counter{Value: &v},
+			}
+		}
+
+		errFamily.Metric = append(errFamily.Metric, generateMetric(float64(count)))
+	}
+
+	if len(errFamily.Metric) == 0 {
+		return nil
+	}
+
+	return errFamily
+}
+
+func metricsHandler(rsp http.ResponseWriter, req *http.Request) {
+
+	mu.Lock()
+	out := globalCache.Metrics
+	if out == nil {
+		out = make([]*dto.MetricFamily, 0)
 	}
+	errFamily := errorCountMetrics(errorCounts)
+	if errFamily != nil {
+		out = append(out, errFamily)
+	}
+	if time.Now().Sub(globalCache.Date) > opts.TTL {
+		// Only serve stale data once. After that we delete it
```

Comment:
This exporter will be hit by N scrapers (usually 2) -- so this would mean only 1 would get the data. Since this is just the expiration mechanism its probably fine.

Reply:
Yeah. It's not the best solution. I just want to prevent stale reads, since if pulling the info is failing or taking too long we probably want to return no data as opposed to stale data.

Reply:
That I'm not sure about; stale data in some ways is better than no data -- assuming we have a metric for how stale the data is.